### PR TITLE
Implement conditional update

### PR DESCRIPTION
**Pull request by mattwiller**
Implementing FHIR's [conditional update](http://hl7.org/fhir/R4/http.html#cond-update) functionality, which allows clients to upsert a resource based on search criteria

### CODE REVIEW

File: packages/core/src/client.ts [hunk 1]
```diff
@@ -1969,6 +1970,31 @@ export class MedplumClient extends EventTarget {
...
 
+  /**
+   * Upsert a resource: update it in place if it exists, otherwise create it.  This is done in a single, transactional
+   * request to guarantee data consistency.
+   * @param resource - The resource to update or create.
+   * @param query - A FHIR search query to uniquely identify the resource if it already exists.
+   * @param options  - Optional fetch options.
+   * @returns The updated/created resource.
+   */
+  async upsert<T extends Resource>(resource: T, query: string, options?: MedplumRequestOptions): Promise<T> {
```
Comment (codyebberson) [inline packages/core/src/client.ts:1981]: @rahul1  - thoughts on function name?

At minimum, I think we should use `upsertResource()` to follow the pattern of `createResource()`, `updateResource()`, etc, and to give us room in the future for other potential upsert methods (??).

I'd also be ok to reuse the existing `createResourceIfNoneExist()` for smoother migration.  Not sure if the API contract is similar enough to reuse though.

File: packages/fhir-router/src/repo.ts [hunk 2]
```diff
@@ -92,6 +94,19 @@ export interface FhirRepository<TClient = unknown> {
...
 
+  /**
+   * Creates or updates a resource based on a search criteria.
+   *
+   * See: https://www.hl7.org/fhir/http.html#cond-update
+   * @param resource - The FHIR resource to create or update.
+   * @param search - The search criteria to find an existing resource to update.
+   * @returns The updated resource.
+   */
+  conditionalUpdate<T extends Resource>(
```
Comment (codyebberson) [inline packages/fhir-router/src/repo.ts:105]: Hm, all resource-oriented operations currently include `Resource` in the function name.

`conditionallyUpdateResource` is probably the most technically accurate but long and an eyesore.

Objections to `upsertResource()`, which would then align with the name in `MedplumClient`?

**Comment by reshmakh** [comment]
Add log line

File: packages/core/src/client.ts [hunk 3]
```diff
@@ -1969,6 +1969,31 @@ export class MedplumClient extends EventTarget {
...
 
+  /**
+   * Upsert a resource: update it in place if it exists, otherwise create it.  This is done in a single, transactional
+   * request to guarantee data consistency.
+   * @param resource - The resource to update or create.
+   * @param query - A FHIR search query to uniquely identify the resource if it already exists.
+   * @param options  - Optional fetch options.
+   * @returns The updated/created resource.
+   */
+  async upsertResource<T extends Resource>(resource: T, query: string, options?: MedplumRequestOptions): Promise<T> {
```
Comment (rahul1) [inline packages/core/src/client.ts:1980]: @mattwiller , for the `query` param, can we support the `query: QueryTypes`, similar to what we use in `searchResources`. 

Being able to pass a dictionary in can be very handy

Comment (codyebberson) [reply packages/core/src/client.ts:1980]: This is a good call.  See `searchResources`.  You can then use `fhirSearchUrl` to generate the URL as well.

```ts
    const url = this.fhirSearchUrl(resourceType, query);
```

**Review by codyebberson — APPROVED** [review]
👍 🎉

**Pull request merged**